### ISSUE TITLE
Type "Foo" cannot be assigned to "Self@Foo" regression in 1.1.335

### ISSUE THREAD

**Issue by ottumm** [microsoft/pyright#6483]
The following worked in 1.1.334 but errors in 1.1.335:

Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.335&pythonVersion=3.12&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDKApgDbABQ5AxiQIYDO9UAYmGAFzlTdQAmRwKAH0hqJDBEAKeqWABKTjyVQZZAHRCARrQBeUALxQAjJSX9B2kNNlyoAWgB8hWYuXcQRGAFcQKFm0k5ciA):

```python
from typing import Self

class Foo:
    def __init__(self):
        self._baz = 1

    def bar(self) -> Self:
        return Foo()

```

```
Expression of type "Foo" cannot be assigned to return type "Self@Foo"
  Type "Foo" cannot be assigned to type "Self@Foo"  (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
This isn't a regression. The previous behavior was incorrect, and this was fixed in a recent version.

You can do one of the following to make the code sample type check without errors:
```python
    def bar(self) -> "Foo":
        return Foo()
```

or

```python
    def bar(self) -> Self:
        return self.__class__()
```

or

```python
    def bar(self) -> Self:
        return type(self)()
```

**Comment by ottumm** [comment]
Ahhhh, thank you, I was misunderstanding the meaning of `Self`.
